op-node/sources/rollupclients.go

```diff
@@ -0,0 +1,94 @@
+package sources
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+)
+
+type RollupClients struct {
+	Rpcs []*RollupClient
+
+	lock    sync.RWMutex
+	highest *RollupClient
+}
+
+func NewRollupClients(rpcs []*RollupClient) *RollupClients {
+	return &RollupClients{
+		Rpcs: rpcs,
+	}
+}
+
+func (r *RollupClients) OutputAtBlock(ctx context.Context, blockNum uint64) (*eth.OutputResponse, error) {
+	r.lock.RLock()
+	rpc := r.highest
+	r.lock.RUnlock()
+
+	if rpc == nil {
+		return nil, errors.New("no RPCs available")
+	}
+
+	return rpc.OutputAtBlock(ctx, blockNum)
+}
+
+func (r *RollupClients) SyncStatus(ctx context.Context) (*eth.SyncStatus, error) {
+	syncStatus := func(ctx context.Context) ([]*eth.SyncStatus, error) {
+		g, ctx := errgroup.WithContext(ctx)
+
+		results := make([]*eth.SyncStatus, len(r.Rpcs))
+		for i, rpc := range r.Rpcs {
+			i, rpc := i, rpc // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				result, err := rpc.SyncStatus(ctx)
+				if err == nil {
+					results[i] = result
+				}
+				return err
+			})
+		}
+		if err := g.Wait(); err != nil {
+			return nil, err
+		}
```

Comment:
If some nodes are unable to access, the `syncStatus` closure will return `(nil, err)`. Currently, `RollupClients.SyncStatus` returns an error in such cases. However, it may be more beneficial to discard these failing nodes and only return the successful results.

Reply:
fix